# AIM HW 10 - MARLA GOODMAN 

### (Task 1 was completed outside the Notebook. I entered the given assignment data into Excel, saved the Excel file as a CSV, and uploaded the CSV file to my Github repository)

### Task 2: Transforming Original Data

Import pandas library:

In [249]:
import pandas as pd

Read csv file into notebook:

In [250]:
df=pd.read_csv('https://raw.githubusercontent.com/MarlaGoodman/AIM-5001/master/hw10data.csv')

In [251]:
#show df
df

Unnamed: 0 Unnamed: 1  Los Angeles  Phoenix  San Diego  San Francisco  \
0     ALASKA    on time        497.0    221.0      212.0          503.0   
1        NaN    delayed         62.0     12.0       20.0          102.0   
2        NaN        NaN          NaN      NaN        NaN            NaN   
3    AM WEST    on time        694.0   4840.0      383.0          320.0   
4        NaN    delayed        117.0    415.0       65.0          129.0   

   Seattle  
0   1841.0  
1    305.0  
2      NaN  
3    201.0  
4     61.0

To clean up the data, I am dropping the middle row that contains only NaNs.

In [252]:
#where the LA column has an 'NaN', drop that whole row
dropped_df=df.dropna(subset = ['Los Angeles'])
dropped_df

Unnamed: 0 Unnamed: 1  Los Angeles  Phoenix  San Diego  San Francisco  \
0     ALASKA    on time        497.0    221.0      212.0          503.0   
1        NaN    delayed         62.0     12.0       20.0          102.0   
3    AM WEST    on time        694.0   4840.0      383.0          320.0   
4        NaN    delayed        117.0    415.0       65.0          129.0   

   Seattle  
0   1841.0  
1    305.0  
3    201.0  
4     61.0

I am now filling in the first NaN with 'ALASKA':

In [253]:
#fill in NA with 'ALASKA', limiting this change only to the first NA
df_edited=dropped_df.fillna('ALASKA', limit=1)

And also filling in the last NaN with 'AM WEST':

In [254]:
#fill in the NA of the df_edited dataframe with 'AM WEST'
data=df_edited.fillna('AM WEST')

In [255]:
#show data
data

Unnamed: 0 Unnamed: 1  Los Angeles  Phoenix  San Diego  San Francisco  \
0     ALASKA    on time        497.0    221.0      212.0          503.0   
1     ALASKA    delayed         62.0     12.0       20.0          102.0   
3    AM WEST    on time        694.0   4840.0      383.0          320.0   
4    AM WEST    delayed        117.0    415.0       65.0          129.0   

   Seattle  
0   1841.0  
1    305.0  
3    201.0  
4     61.0

To make the data look a bit cleaner, I am renaming the columns:

In [256]:
#rename columns of 'data' dataframe
data.columns = ['Airline', 'Arrival','Los Angeles','Phoenix','San Diego','San Francisco','Seattle']
data

Airline  Arrival  Los Angeles  Phoenix  San Diego  San Francisco  Seattle
0   ALASKA  on time        497.0    221.0      212.0          503.0   1841.0
1   ALASKA  delayed         62.0     12.0       20.0          102.0    305.0
3  AM WEST  on time        694.0   4840.0      383.0          320.0    201.0
4  AM WEST  delayed        117.0    415.0       65.0          129.0     61.0

The above dataframe is not tidy. I want each observation to get its own row. An observation is a very specific set of variables pertaining to flights. 

For example, the first observation is 497 planes that are apart of Alaska airline that arrived on time to LA. Therefore, in one row, I want 'Alaska', 'on time', 'LA', and '497'. 

This would then have the new dataframe contain 4 columns: 'Airline','Arrival' (on time arrival or delayed arrival),'Destination', and 'Number of Planes' (# of Planes). I will use the melt function to merge the 5 city columns into one column, which I will name 'Destination'.

In [257]:
#melt the data so it appears on long form
melted=pd.melt(data,['Airline','Arrival'])

In [258]:
#rename columns within melted data
melted.columns=['Airline','Arrival','Destination','# of Planes']

In [259]:
#show melted data
melted

Airline  Arrival    Destination  # of Planes
0    ALASKA  on time    Los Angeles        497.0
1    ALASKA  delayed    Los Angeles         62.0
2   AM WEST  on time    Los Angeles        694.0
3   AM WEST  delayed    Los Angeles        117.0
4    ALASKA  on time        Phoenix        221.0
5    ALASKA  delayed        Phoenix         12.0
6   AM WEST  on time        Phoenix       4840.0
7   AM WEST  delayed        Phoenix        415.0
8    ALASKA  on time      San Diego        212.0
9    ALASKA  delayed      San Diego         20.0
10  AM WEST  on time      San Diego        383.0
11  AM WEST  delayed      San Diego         65.0
12   ALASKA  on time  San Francisco        503.0
13   ALASKA  delayed  San Francisco        102.0
14  AM WEST  on time  San Francisco        320.0
15  AM WEST  delayed  San Francisco        129.0
16   ALASKA  on time        Seattle       1841.0
17   ALASKA  delayed        Seattle        305.0
18  AM WEST  on time        Seattle        201.0
19  AM WEST  delayed        Seattle         61.0

### Task 3: Questions

Question 1. For each city, which airline had the best on time performance?

I want to create a subset from the melted data for each city. Then I'll pull out just the 'on time' rows (via their even indexes).

In [260]:
#creating subsets from melted data
LA=melted[:4]
Ph=melted[4:8]
SD=melted[8:12]
SF=melted[12:16]
Se=melted[16:20]

In [261]:
#use lambda for row locating in subset
#slice subset to select rows with even index labels
LA.iloc[lambda x: x.index % 2 == 0]

Airline  Arrival  Destination  # of Planes
0   ALASKA  on time  Los Angeles        497.0
2  AM WEST  on time  Los Angeles        694.0

AM West has the best on time performance for LA.

In [262]:
#use lambda for row locating in subset
#slice subset to select rows with even index labels
Ph.iloc[lambda x: x.index % 2 == 0]

Airline  Arrival Destination  # of Planes
4   ALASKA  on time     Phoenix        221.0
6  AM WEST  on time     Phoenix       4840.0

AM West has the best on time performance for Phoenix.

In [263]:
#use lambda for row locating in subset
#slice subset to select rows with even index labels
SD.iloc[lambda x: x.index % 2 == 0]

Airline  Arrival Destination  # of Planes
8    ALASKA  on time   San Diego        212.0
10  AM WEST  on time   San Diego        383.0

AM West had the best on time performance for San Diego.

In [264]:
#use lambda for row locating in subset
#slice subset to select rows with even index labels
SF.iloc[lambda x: x.index % 2 == 0]

Airline  Arrival    Destination  # of Planes
12   ALASKA  on time  San Francisco        503.0
14  AM WEST  on time  San Francisco        320.0

Alaska had the best on time performance for San Francisco.

In [265]:
#use lambda for row locating in subset
#slice subset to select rows with even index labels
Se.iloc[lambda x: x.index % 2 == 0]

Airline  Arrival Destination  # of Planes
16   ALASKA  on time     Seattle       1841.0
18  AM WEST  on time     Seattle        201.0

Alaska had the best on time performace for Seattle.

Question 2. Which airline had the best overall on time performance?

First I will get all the on time rows from the melted data using the same lambda even index function. Then I will group the Airlines and use the sum function. 

In [266]:
#pull out even indexes ('on time' rows)
on_time=melted.iloc[lambda x: x.index % 2 == 0]
on_time

Airline  Arrival    Destination  # of Planes
0    ALASKA  on time    Los Angeles        497.0
2   AM WEST  on time    Los Angeles        694.0
4    ALASKA  on time        Phoenix        221.0
6   AM WEST  on time        Phoenix       4840.0
8    ALASKA  on time      San Diego        212.0
10  AM WEST  on time      San Diego        383.0
12   ALASKA  on time  San Francisco        503.0
14  AM WEST  on time  San Francisco        320.0
16   ALASKA  on time        Seattle       1841.0
18  AM WEST  on time        Seattle        201.0

In [267]:
#group by 'Airline' and sum the numbers
on_time.groupby(by='Airline').sum()

# of Planes
Airline             
ALASKA        3274.0
AM WEST       6438.0

AM West airline had the best overall on time performance.

### Task 4: Changes to Wide Format?

When transforming the data back into a 'wide' format, I would not put the information back in the original 'wide' structure. This is because the original one was hard to read:

In [268]:
#original wide format
df

Unnamed: 0 Unnamed: 1  Los Angeles  Phoenix  San Diego  San Francisco  \
0     ALASKA    on time        497.0    221.0      212.0          503.0   
1        NaN    delayed         62.0     12.0       20.0          102.0   
2        NaN        NaN          NaN      NaN        NaN            NaN   
3    AM WEST    on time        694.0   4840.0      383.0          320.0   
4        NaN    delayed        117.0    415.0       65.0          129.0   

   Seattle  
0   1841.0  
1    305.0  
2      NaN  
3    201.0  
4     61.0

Even when I cleaned up the NaNs, it was still hard to read:

In [269]:
#show cleaned up data table
data

Airline  Arrival  Los Angeles  Phoenix  San Diego  San Francisco  Seattle
0   ALASKA  on time        497.0    221.0      212.0          503.0   1841.0
1   ALASKA  delayed         62.0     12.0       20.0          102.0    305.0
3  AM WEST  on time        694.0   4840.0      383.0          320.0    201.0
4  AM WEST  delayed        117.0    415.0       65.0          129.0     61.0

There are many variables: Airline (2), Arrival (2), destinations (5), and values. In the above table, all the varibles are thrown in and left to the reader to make sense of. An organized way to present all the variables is to create dimensions with hierarchal indexing. I will use the pivot function to do so accordingly:

In [270]:
#pivot the data to make use of hierarchal indexing
data.pivot('Arrival','Airline')

Los Angeles         Phoenix         San Diego         San Francisco  \
Airline      ALASKA AM WEST  ALASKA AM WEST    ALASKA AM WEST        ALASKA   
Arrival                                                                       
delayed        62.0   117.0    12.0   415.0      20.0    65.0         102.0   
on time       497.0   694.0   221.0  4840.0     212.0   383.0         503.0   

                Seattle          
Airline AM WEST  ALASKA AM WEST  
Arrival                          
delayed   129.0   305.0    61.0  
on time   320.0  1841.0   201.0

The above chart is much easier to read. For example, I say: "find me the value of delayed arrival flights for Alaska airline heading to LA". The reader works backwards through three easy steps, as follows: She first looks at the top index to find LA. Then she looks to LA's two sub indexes to find Alaska. Finally, she finds the intersection of the LA/Alaska column and the Delayed row. She comes up with the value of 62. 